msal/apps.go

```diff
@@ -0,0 +1,442 @@
+
+package msal
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/AzureAD/microsoft-authentication-library-for-go/internal/msalbase"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/internal/requests"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/internal/storage"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/msal/cache"
+)
+
+type noopCacheAccessor struct{}
+
+func (n noopCacheAccessor) Replace(cache cache.Unmarshaler) {}
+func (n noopCacheAccessor) Export(cache cache.Marshaler)    {}
+
+// manager provides an internal cache. It is defined to allow faking the cache in tests.
+// In all production use it is a *storage.Manager.
+type manager interface {
+	Read(ctx context.Context, authParameters msalbase.AuthParametersInternal, webRequestManager requests.WebRequestManager) (msalbase.StorageTokenResponse, error)
+	Write(authParameters msalbase.AuthParametersInternal, tokenResponse msalbase.TokenResponse) (msalbase.Account, error)
+	GetAllAccounts() ([]msalbase.Account, error)
+}
+
+type clientApplication struct {
+	webRequestManager           requests.WebRequestManager
+	clientApplicationParameters *clientApplicationParameters
+	manager                     manager // *storage.Manager or fakeManager in tests
+	cacheAccessor               cache.ExportReplace
+}
+
+func createClientApplication(httpClient HTTPClient, clientID string, authority string) (*clientApplication, error) {
+	params, err := createClientApplicationParameters(clientID, authority)
+	if err != nil {
+		return nil, err
+	}
+
+	return &clientApplication{
+		webRequestManager:           createWebRequestManager(httpClient),
+		clientApplicationParameters: params,
+		cacheAccessor:               noopCacheAccessor{},
+		manager:                     storage.New(),
+	}, nil
+}
+
+func (client *clientApplication) createAuthCodeURL(ctx context.Context, authCodeURLParameters AuthorizationCodeURLParameters) (string, error) {
+	return authCodeURLParameters.createURL(ctx, client.webRequestManager, client.clientApplicationParameters.createAuthenticationParameters())
+}
+
+func (client *clientApplication) acquireTokenSilent(ctx context.Context, silent AcquireTokenSilentParameters) (msalbase.AuthenticationResult, error) {
+	authParams := client.clientApplicationParameters.createAuthenticationParameters()
+	silent.augmentAuthenticationParameters(&authParams)
+
+	// TODO(jdoak): Think about removing this after refactor.
+	if s, ok := client.manager.(cache.Serializer); ok {
+		client.cacheAccessor.Replace(s)
+		defer client.cacheAccessor.Export(s)
+	}
+
+	storageTokenResponse, err := client.manager.Read(ctx, authParams, client.webRequestManager)
+	if err != nil {
+		return msalbase.AuthenticationResult{}, err
+	}
+
+	result, err := msalbase.CreateAuthenticationResultFromStorageTokenResponse(storageTokenResponse)
+	if err != nil {
+		if reflect.ValueOf(storageTokenResponse.RefreshToken).IsNil() {
```

Comment:
we missed this in a previous refactor, this doesn't make much sense anymore and should be an IsZero check. can we get a todo or a issue?